From 306c1985b0f8a6f07564bcd422f0f376b379f9b3 Mon Sep 17 00:00:00 2001
From: Ang Tien Sung <tien.sung.ang@intel.com>
Date: Tue, 23 Nov 2021 11:01:25 +0800
Subject: [PATCH 16/20] HSD #14015013554: fpga: stratix10-soc: Support
 multi-client multi-threading

commit bef4c91328840e278926d3c667890ed1d805020b from
https://github.com/altera-opensource/linux-socfpga/commits/socfpga-5.4.124-lts

This minor fix is to ensure, that the FPGA driver releases the SVC mailbox
resource after usage.
The change also ensures, that the FPGA driver is able to function properly
when run together with another SVC driver like soc64-hwmon and intel_fcs.

Signed-off-by: Ang Tien Sung <tien.sung.ang@intel.com>
Integrated-by: Abhishek Paliwal <paliwal.abhishek@windriver.com>
---
 drivers/fpga/stratix10-soc.c | 16 +++++++++++-----
 1 file changed, 11 insertions(+), 5 deletions(-)

diff --git a/drivers/fpga/stratix10-soc.c b/drivers/fpga/stratix10-soc.c
index ce58fd30bf17..e2de802030ee 100644
--- a/drivers/fpga/stratix10-soc.c
+++ b/drivers/fpga/stratix10-soc.c
@@ -235,20 +235,20 @@ static int s10_ops_write_init(struct fpga_manager *mgr,
 			       &ctype, sizeof(ctype),
 			       s10_receive_callback);
 	if (ret < 0)
-		goto init_done;
+		goto init_error;
 
 	ret = wait_for_completion_timeout(
 		&priv->status_return_completion, S10_RECONFIG_TIMEOUT);
 	if (!ret) {
 		dev_err(dev, "timeout waiting for RECONFIG_REQUEST\n");
 		ret = -ETIMEDOUT;
-		goto init_done;
+		goto init_error;
 	}
 
 	ret = 0;
 	if (!test_and_clear_bit(SVC_STATUS_OK, &priv->status)) {
 		ret = -ETIMEDOUT;
-		goto init_done;
+		goto init_error;
 	}
 
 	/* Allocate buffers from the service layer's pool. */
@@ -257,15 +257,18 @@ static int s10_ops_write_init(struct fpga_manager *mgr,
 		if (!kbuf) {
 			s10_free_buffers(mgr);
 			ret = -ENOMEM;
-			goto init_done;
+			goto init_error;
 		}
 
 		priv->svc_bufs[i].buf = kbuf;
 		priv->svc_bufs[i].lock = 0;
 	}
 
-init_done:
+	goto init_done;
+
+init_error:
 	stratix10_svc_done(priv->chan);
+init_done:
 	return ret;
 }
 
@@ -376,6 +379,9 @@ static int s10_ops_write(struct fpga_manager *mgr, const char *buf,
 	if (!s10_free_buffers(mgr))
 		dev_err(dev, "%s not all buffers were freed\n", __func__);
 
+	if (ret < 0)
+		stratix10_svc_done(priv->chan);
+
 	return ret;
 }
 
-- 
2.31.1

